File: leet_BTREE_0515_max_of_each___BFS.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import (
    "container/list"
    "math"
)

func largestValues(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }

    D := list.New()
    D.PushBack(root)
    var res []int

    for D.Len() > 0 {
        len := D.Len()
        temp := math.MinInt32

        for i := 0; i < len; i++ {
            node := D.Remove(D.Front()).(*TreeNode)
            temp = int(math.Max(float64(temp), float64(node.Val)))

            if node.Left != nil {
                D.PushBack(node.Left)
            }
            if node.Right != nil {
                D.PushBack(node.Right)
            }
        }

        res = append(res, temp)
    }

    return res
}

File: leet_BTREE_0515_max_of_each___BFS.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function largestValues(root: TreeNode | null): number[] {
    if (root === null)
        return []
    let D: TreeNode[] = [root]
    let res: number[] = []
    while (D.length) {
        let len:number = D.length
        let temp:number = Number.MIN_SAFE_INTEGER
        while (len--) {
            let node: TreeNode = D.shift()
            temp = Math.max(temp, node.val)
            if (node.left)
                D.push(node.left)
            if (node.right)
                D.push(node.right)
        }
        res.push(temp)
    }
    return res
};
